lib/vector-core/src/transform/mod.rs

```diff
@@ -159,54 +160,73 @@ where
     T: FunctionTransform,
 {
     fn transform(&mut self, event: Event, output: &mut TransformOutputsBuf) {
-        FunctionTransform::transform(self, &mut output.primary_buffer, event);
+        FunctionTransform::transform(
+            self,
+            output.primary_buffer.as_mut().expect("no default output"),
+            event,
+        );
     }
 }
 
 // TODO: this is a bit ugly when we already have the above impl
 impl SyncTransform for Box<dyn FunctionTransform> {
     fn transform(&mut self, event: Event, output: &mut TransformOutputsBuf) {
-        FunctionTransform::transform(self.as_mut(), &mut output.primary_buffer, event);
+        FunctionTransform::transform(
+            self.as_mut(),
+            output.primary_buffer.as_mut().expect("no default output"),
+            event,
+        );
     }
 }
 
 pub struct TransformOutputs {
-    primary_output: Fanout,
+    outputs_spec: Vec<Output>,
+    primary_output: Option<Fanout>,
     named_outputs: HashMap<String, Fanout>,
 }
 
 impl TransformOutputs {
-    pub fn new(
-        named_outputs_in: Vec<String>,
-    ) -> (Self, HashMap<Option<String>, fanout::ControlChannel>) {
+    pub fn new(outputs_in: Vec<Output>) -> (Self, HashMap<Option<String>, fanout::ControlChannel>) {
+        let outputs_spec = outputs_in.clone();
```

Comment:
I've a doubt here, this is being run each time the topology is built, not for every event or batch of event, right?

Reply:
Yes, this is built once when the transform itself is built.